android/src/main/java/com/ryanheise/audioservice/AudioServicePlugin.java

```diff
@@ -129,24 +133,43 @@ public void onDetachedFromEngine(FlutterPluginBinding binding) {
 
 	@Override
 	public void onAttachedToActivity(ActivityPluginBinding binding) {
-		clientHandler = new ClientHandler(flutterPluginBinding.getFlutterEngine().getDartExecutor());
+		activityPluginBinding = binding;
+		clientHandler = new ClientHandler(flutterPluginBinding.getBinaryMessenger());
 		clientHandler.activity = binding.getActivity();
+		registerOnNewIntentListener();
 	}
 
 	@Override
 	public void onDetachedFromActivityForConfigChanges() {
+		activityPluginBinding.removeOnNewIntentListener(newIntentListener);
 	}
 
 	@Override
 	public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
+		activityPluginBinding = binding;
 		clientHandler.activity = binding.getActivity();
+		registerOnNewIntentListener();
 	}
 
 	@Override
 	public void onDetachedFromActivity() {
+		activityPluginBinding.removeOnNewIntentListener(newIntentListener);
+		newIntentListener = null;
+		activityPluginBinding = null;
 		clientHandler = null;
 	}
 
+	private void registerOnNewIntentListener() {
+		activityPluginBinding.addOnNewIntentListener(newIntentListener = new NewIntentListener() {
+			@Override
+			public boolean onNewIntent(Intent intent) {
+				clientHandler.activity.setIntent(new Intent().setAction(intent.getAction()));
+				clientHandler.invokeMethod("notificationClicked", intent.getAction() == null);
```

Comment:
Is `intent.getAction() == null` a safe way to determine that this intent is ours? I think the problem is the pending intent I set on the notification with `setContentIntent` which is the default one that doesn't have any action on it, but perhaps we could pass in a unique action that would not be confused with anything else.

Reply:
Yes we can pass unique action but I do not know how to do so with the current code. Here is what I think can be done, but `MainActivity.class` i.e the result activity to be opened is not accessible. Should it be replaced with `FlutterActivity.class`?
```java
private PendingIntent getPendingIntent() {
   Intent notifyIntent = new Intent(this, MainActivity.class); // Replace with FlutterActivity.class
   // Set the Activity to start in a new, empty task
   notifyIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
   // Set the actions
   notifyIntent.setAction("audio_service_action");
   // Create the PendingIntent
   return PendingIntent.getActivity(this, 0, notifyIntent, PendingIntent.FLAG_UPDATE_CURRENT);
}
```
As I am not sure what must be done here I let the pendingIntent remain untouched. Also is there a way to pass an action to the existing pendingIntent?

Reply:
I had forgotten, but in `AudioService.init` there is this line:

```java
contentIntent = PendingIntent.getActivity(context, REQUEST_CONTENT_INTENT, intent, PendingIntent.FLAG_UPDATE_CURRENT);
```

This intent is the one that is set on the media session which is later used by `mediaSession.getController().getSessionActivity()` (if my understanding is correct). So that code above is where I believe the action could be set on the intent.

Reply:
Awesome. I had not seen this. We can set the action here. What would you like the action be named?